### DISCUSSION TITLE
Animate particle path calculation results

### DISCUSSION THREAD

**Discussion by apcode** [pyvista/pyvista#4945] [Q&A]
I can't for the life of me work out how to do this with pyvista:

I have an other library that calculates particle paths in a field over time. 
Starting from a point for each particle, it's next position is calculated every second.

I am displaying these particle paths as polydata lines.

I would like to view these paths as a polydata updates over time in an animation - onscreen, not in an offscreen video.

Any help on how to achieve this?

**Comment by banesullivan** [comment]
Here are some leads for you:

- https://github.com/pyvista/pyvista/discussions/4607#discussioncomment-6318790
- https://snawarhussain.com/blog/data%20visualization/python/tutorial/3D-motion-capture-data-animation-with-pyvista/

If you want this on screen, the you'll need to use something like pyvistaqt: https://qtdocs.pyvista.org/

**Comment by akaszynski** [comment]
Year late, but I needed to somehow animate streamlines of a CFD solution, which isn't really possible if it's just a single time solution, so I opted to animate particles.

https://github.com/user-attachments/assets/8819dd52-f7ac-454d-9382-cb8cdb20c844

The trick is to take the output of streamlines:

```py
bc_points = pv.PointSet(benzene.points)
b_sl = benzene.streamlines_from_source(
    bc_points, vectors=VEL_NAME, max_time=100, integration_direction="forward"
)
```
Then, use [contour](https://docs.pyvista.org/api/core/_autosummary/pyvista.datasetfilters.contour#pyvista.DataSetFilters.contour) to evaluate the `IntegrationTime` of the streamlines to create individual points. I rounded here to ensure I had `n_steps` unique time points.

```py
# contour and compute contour centers
print(f"Contouring {n_steps} steps...", end="", flush=True)
b_pts = b_sl.contour(n_steps, "IntegrationTime")
b_itime = np.round(b_pts["IntegrationTime"], 5)
b_pts["itime"] = b_itime
b_sc = np.unique(b_itime)
assert b_sc.size == n_steps - 1
```

Finally, create (and then remove) the points for each time step:

```py
                for ii in tqdm(rng, desc="Generating animation"):
                    # add benzene "particles"
                    b_epoints = b_pts.extract_points(
                        np.isin(b_pts["itime"], b_sc[ii::n_splits])
                    )
                    b_actor = pl.add_mesh(
                        b_epoints,
                        render_points_as_spheres=True,
                        show_scalar_bar=False,
                        scalars="TEMPERATURE",
                        cmap="reds",
                        rng=b_rng,
                        point_size=point_size,
                        ambient=ambient,
                        render=False,
                    )
                #...
                pl.write_frame()
                pl.remove_actor(b_actor, render=False)
```

There'a lot more to generate the streamlines and figures, but hopefully someone else finds this helpful as a starting point.

**Reply by MatthewFlamm** [reply]
Thanks for sharing.  I wonder if this should be a built in filter in PyVista to generate objects along streamlines at constant (or custom) intervals?

`contour` for interpolation of the streamlines in time is a nice PyVista use that isn't immediately obvious.  I have used `np.interp` for this purpose in the past, which makes it slightly easier to handle end point discrepancies in my opinion.

Also a breadcrumb note, since you used `max_time`.  This parameter is a bit weird, and we probably need to rename it in PyVista.  See https://github.com/pyvista/pyvista/issues/2301.

**Reply by akaszynski** [reply]
I think it would be great to implement this in PyVista. I'll have to try out `np.interp`.

**Reply by MatthewFlamm** [reply]
I'd love for `vtkStreamTracer` or PyVista's `streamlines(_from_source)` to integrate all the way to the edge of a domain, but this is best done in vtk IMO.

**Reply by MatthewFlamm** [reply]
> I'd love for vtkStreamTracer or PyVista's streamlines(_from_source) to integrate all the way to the edge of a domain, but this is best done in vtk IMO.

I am about 50% sure that we can use `vtkLagrangianParticleTracker` to do this. As it is meant to track particles of physical sizes, we would need to test it with particles of 0 size and to set the initial velocity condition to be the same as the velocity field to generate a streamline.  It isn't a drop-in replacement, but it might be useful to base a filter off `vtkLagrangianParticleTracker` and contrast it with `vtkStreamTracer`.

**Reply by MatthewFlamm** [reply]
> > I'd love for vtkStreamTracer or PyVista's streamlines(_from_source) to integrate all the way to the edge of a domain, but this is best done in vtk IMO.
> 
> I am about 50% sure that we can use `vtkLagrangianParticleTracker` to do this. As it is meant to track particles of physical sizes, we would need to test it with particles of 0 size and to set the initial velocity condition to be the same as the velocity field to generate a streamline. It isn't a drop-in replacement, but it might be useful to base a filter off `vtkLagrangianParticleTracker` and contrast it with `vtkStreamTracer`.

The default algorithm used by `vtkLagrangianParticleTracker` does not handle well very small diameters.  It becomes unstable when using an explicit ODE solver in vtk.  vtk does not have any built-in implicit ODE solvers. At very small diameters it requires very small integration time steps.  I still think it is a useful addition to PyVista and will work on it when I get time.
